### DISCUSSION TITLE
Nulleinspeisung mit Shelly 3EM

### DISCUSSION THREAD

**Discussion by zanycam** [tbnobody/OpenDTU#567] [Ideas]
Hallo, wäre es möglich es umzusetzen das OpenDTU in Zusammenspiel mit einem Shelly 3EM ohne weitere Hard/Software eine Nulleinspeisung (auch über mehrere WR) realisiert? Per http-Abfrage liefert der Shelly ein json-File mit dem aktuellen saldierten Netzbezug zurück:

http://<3EM_IP_ADDRESS>/status?emeters

{"wifi_sta":{"connected":true,"ssid":"<SSID>","ip":"<IP_Adresse>","rssi":-75},"cloud":{"enabled":true,"connected":true},"mqtt":{"connected":false},"time":"18:15","unixtime":1675703713,"serial":32281,"has_update":false,"mac":"C45BBE7973E1","cfg_changed_cnt":0,"actions_stats":{"skipped":0},"relays":[{"ison":false,"has_timer":false,"timer_started":0,"timer_duration":0,"timer_remaining":0,"overpower":false,"is_valid":true,"source":"input"}],"emeters":[{"power":53.92,"pf":0.35,"current":0.66,"voltage":234.66,"is_valid":true,"total":315959.1,"total_returned":964.3},{"power":774.51,"pf":0.95,"current":3.48,"voltage":234.25,"is_valid":true,"total":345884.4,"total_returned":17.4},{"power":54.50,"pf":0.58,"current":0.40,"voltage":234.87,"is_valid":true,"total":143049.4,"total_returned":7027.0}],"total_power":882.93,"emeter_n":{"current":0.00,"ixsum":2.83,"mismatch":false,"is_valid":false},"fs_mounted":true,"v_data":1,"ct_calst":0,"update":{"status":"idle","has_update":false,"new_version":"20221027-110030/v1.12.1-ga9117d3","old_version":"20221027-110030/v1.12.1-ga9117d3"},"ram_total":49928,"ram_free":31692,"fs_size":233681,"fs_free":155118,"uptime":1049255}

wobei „total_power“ der aktuelle Netzbezug in Watt ist.

Viele Grüße,
Sandy

**Comment by stefan123t** [comment]
Wir haben so einen ähnlichen Feature Request etwas Allgemeiner gehalten hier https://github.com/lumapu/ahoy/issues/606

Prinzipiell sollte die Abfrage eines REST API Endpoints und je ein JSON JQ-Filter für Verbrauch und Einspeisung hierfür genügen.
Das erste zu unterstützende Target wäre m.E. ein OpenSource TASMOTA SML Device wie z.B. der Hichi Lesekopf.
Aber das Prinzip lässt sich auch analog für andere Endgeräte wie ein Shelly 3EM+ oder irgendwelche anderen WLAN-fähigen Strommessgeräte im selben Netz wie die OpenDTU / AhoyDTU nutzen.

**Comment by FrodoVDR** [comment]
Das dürfte aber problematisch werden, der neue Shelly Pro EM3, hat wieder eine andere API und wird wie folgt abgefragt:
# wget -q -O - http://IP-ShellyPro3EM/rpc/EM.GetStatus?id=0
{"id":0,"a_current":2.780,"a_voltage":232.6,"a_act_power":-547.9,"a_aprt_power":646.2,"a_pf":-0.87,"b_current":3.765,"b_voltage":232.9,"b_act_power":502.3,"b_aprt_power":876.9,"b_pf":-0.70,"c_current":0.384,"c_voltage":233.9,"c_act_power":63.4,"c_aprt_power":89.7,"c_pf":-0.78,"n_current":null,"total_current":6.929,"total_act_power":17.748,"total_aprt_power":1612.833, "user_calibrated_phase":[]}
# wget -q -O - http://IP-ShellyPro3EM/rpc/EMdata.GetStatus?id=0
{"id":0,"a_total_act_energy":11185.16,"a_total_act_ret_energy":9029.99,"b_total_act_energy":43724.67,"b_total_act_ret_energy":7.81,"c_total_act_energy":11231.15,"c_total_act_ret_energy":0.00,"total_act":66140.97, "total_act_ret":9037.80}

**Comment by berni2288** [comment]
Ich versuche das gerade zu implementieren in OpenDTU-OnBattery: https://github.com/helgeerbe/OpenDTU-OnBattery/discussions/141

**Reply by berni2288** [reply]
Wurde fertig implementiert und ist im development branch.

**Reply by stefan123t** [reply]
@zanycam  @FrodoVDR 
please double check the **Dynamic Power Limiter (DPL)** function in the daughter project[ OpenDTU-OnBattery](https://github.com/helgeerbe/OpenDTU-OnBattery/).

I have described a [possible PM+DPL setup](https://github.com/tbnobody/OpenDTU/issues/272#issuecomment-2219644352) with a Shelly Pro3EM as **Power Meter (PM)** here in issue #272.

**Comment by fellpower** [comment]
Wäre cool, wenn das ginge. Hab auchn Shelly 3EM - und fände das Spitze, wenn ich den zum steuern der Einspeisung (oder deren Vermeidung) nehmen könnte ^^

**Reply by virtualpinguin-git** [reply]
Hallo,
das gibt es doch schon:
https://github.com/Selbstbau-PV/Selbstbau-PV-Hoymiles-nulleinspeisung-mit-OpenDTU-und-Shelly3EM
Bei mir läuft es.
Gerd

**Reply by T3rm1** [reply]
Warum sollte man nicht einspeisen wollen?

**Reply by fellpower** [reply]
> Hallo, das gibt es doch schon: https://github.com/Selbstbau-PV/Selbstbau-PV-Hoymiles-nulleinspeisung-mit-OpenDTU-und-Shelly3EM Bei mir läuft es. Gerd

Sinnlos, da es nicht aufm dem ESP läuft - und darum gehts ja - nicht noch 200 Geräte zu benutzen, die Strom fressen ^^

**Reply by bf4ed** [reply]
> Warum sollte man nicht einspeisen wollen?

Frag nicht wieso jemand etwas braucht, und das es sinnlos ist. Es gibt verschiedene Anwendungen.

Bei mir währe es sinnvoll das ich Nulleinspeisung hab, das ich mit Überschuß eine Batterie auflade und in der Nacht mit einem zweiten WR nur die benötigte Menge einspeise. Das ich die Batterie nicht um sonst entlade. Deswegen brauche ich die  Nulleinspeisung.

Und ich hab das mit Shelly 3EM und RaspberryPi probiert und es funktioniert irgendwie nicht.

Es währe besser das alles auf einem ESP32 mit OpenDTU läuft  und er nimmt direkt die Daten von Shelly. Das wäre ideal.

**Reply by virtualpinguin-git** [reply]
Hallo bf4ed,
das kennst du auch schon?
https://github.com/helgeerbe/OpenDTU-OnBattery

**Reply by bf4ed** [reply]
> Hallo bf4ed, das kennst du auch schon? https://github.com/helgeerbe/OpenDTU-OnBattery

Ja, danke. Hab ich später gefunden. Und ausprobiert. Es funktioniert noch nicht, sind viele Einstellungen. Jetzt hab ich aber keine Zeit damit zu spielen. Am Wochenende probiere ich es nochmals. Ich glaube eine Einstellung hab ich falsch gemacht, hab ich erst heute gesehen auf helgeerbe thread.

Währe gut daß sie die Nulleinspeisung hier implementieren, ich mag keine forks.

**Reply by T3rm1** [reply]
> Frag nicht wieso jemand etwas braucht, und das es sinnlos ist. Es gibt verschiedene Anwendungen.

Ich frage, was ich will. Hier ist bisschen Weiterbildung für dich: http://www.wäre-währe.de/

**Reply by bf4ed** [reply]
Mit oder ohne h ist mir doch egal, brauch nicht zu Klugscheißen gegen einem Ausländer.
